nvflare/fuel/f3/drivers/aio_grpc_driver.py

```diff
@@ -0,0 +1,412 @@
+# Copyright (c) 2021-2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import asyncio
+import logging
+import threading
+import time
+import traceback
+from typing import Any, Dict, List
+
+import grpc
+
+from nvflare.fuel.f3.comm_config import CommConfigurator
+from nvflare.fuel.f3.comm_error import CommError
+from nvflare.fuel.f3.connection import BytesAlike, Connection
+from nvflare.fuel.f3.drivers.aio_context import AioContext
+from nvflare.fuel.f3.drivers.driver import ConnectorInfo
+from nvflare.fuel.f3.drivers.grpc.streamer_pb2_grpc import (
+    StreamerServicer,
+    StreamerStub,
+    add_StreamerServicer_to_server,
+)
+
+from .base_driver import BaseDriver
+from .driver_params import DriverCap, DriverParams
+from .grpc.streamer_pb2 import Frame
+from .net_utils import get_address, get_tcp_urls, ssl_required
+
+MAX_MSG_SIZE = 1024 * 1024 * 1024  # 1G
+
+GRPC_DEFAULT_OPTIONS = [
+    ("grpc.max_send_message_length", MAX_MSG_SIZE),
+    ("grpc.max_receive_message_length", MAX_MSG_SIZE),
+]
+
+
+class _ConnCtx:
+    def __init__(self):
+        self.conn = None
+        self.error = None
+        self.waiter = threading.Event()
+
+
+class AioStreamSession(Connection):
+
+    seq_num = 0
+
+    def __init__(
+        self, side: str, aio_ctx: AioContext, connector: ConnectorInfo, conn_props: dict, context=None, channel=None
+    ):
+        super().__init__(connector)
+        self.side = side
+        self.aio_ctx = aio_ctx
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+        self.logger.debug(f"{side}: creating asyncio.Queue")
+        self.oq = asyncio.Queue(16)
```

Comment:
Does oq means output_queue?
Do we want to make this number 16 configurable?

Reply:
Yes. It's Output Queue. We will make this configurable later with all other drivers.
The small number is to create back pressure when the connection can't handle the frame rate. If 16 frames are pending in the queue, the send call will block.